### PR TITLE
feat: update WrapperArray to extend core Array (fix #354)

### LINKED ISSUE #354: Remove WrapperArray
### What problem does this feature solve?
IMHO,
Since WrapperArray has only methods which is the small wrapper method of Array,
it is not absolutly necessary.
I think that the Array of Wrapper is easier to understand than WrapperArray.
I do not think the methods of WrapperArray are often used.
By removing WrapperArray, it become simple and small maintenance cost.

### What does the proposed API look like?
findAll() returns Array of Wrapper and WrapperArray is removed.

### PR DESCRIPTION
**Pull request by lusarz**
In reference to [great suggestion from @38elements ](https://github.com/vuejs/vue-test-utils/issues/354#issuecomment-360433941) I prepared proof of concept of this solution.

I think it works nice, but a few issues I met convinced me that it would be better to remove WrapperArray.

### CODE REVIEW

File: packages/test-utils/src/wrapper-array.js [hunk 1]
```diff
@@ -38,21 +31,17 @@ export default class WrapperArray implements BaseWrapper {
...
-    return this.length > 0 && this.wrappers.every(wrapper => wrapper.exists())
-  }
-
-  filter (predicate: Function): WrapperArray {
-    return new WrapperArray(this.wrappers.filter(predicate))
+    return this.length > 0 && this.every(wrapper => wrapper.exists())
   }
 
   visible (): boolean {
     this.throwErrorIfWrappersIsEmpty('visible')
```
Comment (lusarz) [inline packages/test-utils/src/wrapper-array.js:41]: Personally I think we can remove `throwErrorIfWrappersIsEmpty`, and just return false when `length === 0`. Just like: `return this.length > 0 && this.every(wrapper => wrapper.visible())`

File: test/specs/wrapper/findAll.spec.js [hunk 2]
```diff
@@ -275,9 +273,7 @@ describeWithShallowAndMount('findAll', (mountingMethod) => {
 
   it('returns VueWrapper with length 0 if no nodes matching the ref in options object are found', () => {
     const wrapper = mountingMethod(Component)
-    const preArray = wrapper.findAll({ ref: 'foo' })
-    expect(preArray.length).to.equal(0)
-    expect(preArray.wrappers).to.deep.equal([])
+    expect(wrapper.findAll({ ref: 'foo' })).to.have.lengthOf(0)
```
Comment (lusarz) [inline test/specs/wrapper/findAll.spec.js:276]: Without `WrapperArray extend Array` it was not possible to use `to.have.lengthOf` from chai.

File: packages/test-utils/src/wrapper-array.js [hunk 3]
```diff
@@ -38,21 +30,17 @@ export default class WrapperArray implements BaseWrapper {
...
 
-    return this.wrappers.every(wrapper => wrapper.contains(selector))
+    return this.every(wrapper => wrapper.contains(selector))
   }
 
   exists (): boolean {
-    return this.length > 0 && this.wrappers.every(wrapper => wrapper.exists())
-  }
-
-  filter (predicate: Function): WrapperArray {
```
[the remark this answers is not in the document]
Comment (lusarz) [reply packages/test-utils/src/wrapper-array.js]: Hmm, but now I'm not sure whether it returns WrapperArray or Array instance ...

[the remark this answers is not in the document]
Comment (lusarz) [reply packages/test-utils/src/wrapper-array.js]: Alright, finally reverted this method, and added test that checks if value returned from `filter` method is instance of `WrapperArray`

File: packages/test-utils/src/wrapper-array.js [hunk 4]
```diff
@@ -70,25 +62,25 @@ export default class WrapperArray implements BaseWrapper {
...
 
-    return this.wrappers.every(wrapper => wrapper.hasAttribute(attribute, value))
+    return this.every(wrapper => wrapper.hasAttribute(attribute, value))
   }
 
   hasClass (className: string): boolean {
     this.throwErrorIfWrappersIsEmpty('hasClass')
 
-    return this.wrappers.every(wrapper => wrapper.hasClass(className))
+    return this.every(wrapper => wrapper.hasClass(className))
```
Comment (lusarz) [inline packages/test-utils/src/wrapper-array.js:67]: Not part of PR, but I think that developers may be a confused => when use `hasClass` from `WrapperArray` someone may expect that **any** of children has specified class:

    this.some(wrapper => wrapper.hasClass(className))

Same with most of other methods using `every`

File: packages/test-utils/src/wrapper-array.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (lusarz) [reply packages/test-utils/src/wrapper-array.js]: But unfortunately it cause `flow` check fails :(

**Comment by eddyerburgh** [comment]
What are the issues the convinced you it's better to remove WrapperArray?

**Comment by lusarz** [comment]
I think `Array` has more known and comprehensive API, and contains methods like `some`, `forEach`, `filter`, `map`, `splice` out of box. We are not able to add them all to WrapperArray.

[Here](https://github.com/vuejs/vue-test-utils/issues/354#issuecomment-384589430) I described two more things:
- `every` vs `some`
- chaining of methods

**Comment by eddyerburgh** [comment]
It's worth noting that users can access an array from a WrapperArray with the wrappers property:

```
const divs = wrapper.findAll('div')
divs.wrappers // Array of Wrappers
```

I'm against returning an array from findAll. We had that API in avoriaz and it led to difficult to debug tests.

If anything we could refactor a Wrapper into a collection that defaults to the first item, but I don't think we should do that either.

**Comment by lusarz** [comment]
`wrappers` seems to be some kind of private data of `WrapperArray`, and I think they shouldn't be accessed directly when we have `WrapperArray`. 

I understand what you mean here:

     findAll('div').at(1).props() // throws error [vue-test-utils]: no item exists at 1

it may be helpful for debugging, but from other site I think it's developer fault to try access out of bounds element of array. JS arrays are known to not throwing index-out-of-bounds exception, and developers should be aware of that.

Please consider also:
    
    wrapper.find('div').classes[1]
 
it seems to be similar situation, but there is no `ClassArray` class in `vue-test-utils`.

**Comment by 38elements** [comment]
I accounted the explict benefit.
Since everyone can understand it,
It is not necessary to proof it.
This pull request is unnecessary.
Since #354 has the `discussion` label,
I think this pull request is annoying.

**Comment by 38elements** [comment]
There are same contents which is in #354.
There are two places to discuss about same issue.
This state is inconvenience.
I think that it is better to close this.

**Pull request closed without merging**